app/code/Magento/Email/Model/Template/Filter.php

```
@@ -973,6 +973,9 @@ public function applyInlineCss($html)
         );
         $cssToInline = $this->getCssProcessor()->process($cssToInline);
 
+        $debugMode = $this->_appState->getMode() === \Magento\Framework\App\State::MODE_DEVELOPER;
+        $this->cssInliner->setDebugMode($debugMode);
```

Why do we need this?

---

Emogrifier 1.2 always threw an exception when it encountered something it could not parse (if it didn't directly bail with an XPath error). The filter then either caught and discarded the exception (in production mode) or added an error message (in non-production mode).

Emogrifier 2.0 now has a separate debug mode. In debug mode, it still throws exceptions. In the non-debug mode (the default), it silently ignores anything it cannot parse, and tries to cope the best it can.

So to have both old filter behaviors, we need to enable/disable the Emogrifier debug mode depending on the environment.

---

Got it, thanks for detailed explanation.